### PR TITLE
Add  com.github.emmanueltouzery.cigale

### PR DESCRIPTION
**Pull request by emmanueltouzery**
Hi, I would like to distribute the 'cigale' rust gtk application, of which I'm the author, on flathub.
This is a first shot at the PR to add it, but it still has at least one issue.

I've tried to follow this blog entry as a guide:
https://belmoussaoui.com/2020/02/08/how-to-flatpak-a-rust-application/

But I didn't use meson or ninja. It still think that should be fine. What I did is set up a `update-cargo-sources.sh` script, that i can run before releasing a new version on flathub, that populates `cargo-sources.json`, that contains the source for all the rust dependencies of the application which are needed for offline building.

If I run:

    flatpak-builder --install repo com.github.emmanueltouzery.cigale.json --force-clean --user

With this PR, it fails due to:

```
Running: cargo --offline fetch --manifest-path Cargo.toml --verbose
error: unable to fetch registry `https://github.com/rust-lang/crates.io-index` in offline mode
Try running without the offline flag, or try running `cargo fetch` within your project directory before going offline.
Error: module cigale: Child process exited with code 101
```

It's easy to fix this, I just go in the folder where flatpak builds, for instance `.flatpak-builder/build/cigale-1`, then run:

    cp -R ../../../.cargo .

to copy the .cargo folder from the root. After that the build proceeds successfully. So I'd need to make sure that this .cargo folder is copied in the guest for building, and I'm not sure how to achieve that.

Besides this issue this PR needs to go through the normal review process as well.

### CODE REVIEW

**Comment by bilelmoussaoui** [comment]
bot, build  com.github.emmanueltouzery.cigale

**Comment by bilelmoussaoui** [comment]
bot, build com.github.emmanueltouzery.cigale

**Comment by emmanueltouzery** [comment]
yes, I expect the build to fail, I explained in my original message why. I just don't know how to tell flatpak to include the .cargo/config in the build. That configuration file tells cargo to use the vendor subfolders to enable offline build.

**Comment by emmanueltouzery** [comment]
bot, build com.github.emmanueltouzery.cigale

**Comment by emmanueltouzery** [comment]
ok, now i think this should work. Found out how I can include individual files and folders.

File: com.github.emmanueltouzery.cigale.desktop [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[Desktop Entry]
+Comment=
```
Comment (bilelmoussaoui) [inline com.github.emmanueltouzery.cigale.desktop]:
```suggestion
Comment=
```
I guess you forgot a comment here

File: com.github.emmanueltouzery.cigale.metainfo.xml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bilelmoussaoui) [inline com.github.emmanueltouzery.cigale.metainfo.xml]: The metainfo, desktop file and the icon should be upstreamed.

File: com.github.emmanueltouzery.cigale.json [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        "cargo --offline build --release --verbose",
+        "install -Dm755 ./target/release/cigale -t /app/bin/",
+        "install -Dm644 ./${FLATPAK_ID}.metainfo.xml -t /app/share/metainfo/",
+        "install -Dm644 ./${FLATPAK_ID}.desktop -t /app/share/applications/",
+        "install -Dm644 ./${FLATPAK_ID}.svg -t /app/share/icons/hicolor/scalable/apps/"
+      ],
+      "sources": [
+        {
+          "type": "git",
+          "url": "https://github.com/emmanueltouzery/cigale/",
```
Comment (bilelmoussaoui) [inline com.github.emmanueltouzery.cigale.json]:
```suggestion
          "url": "https://github.com/emmanueltouzery/cigale.git",
```

File: com.github.emmanueltouzery.cigale.metainfo.xml (hunk 2 above)
Comment (emmanueltouzery) [reply com.github.emmanueltouzery.cigale.metainfo.xml]: Upstreamed, you mean put in the Cigale repo itself? In fact they're already there (although the metainfo there is an old version that enables local build). I'll remove them from this repo and use those from the Cigale repo.

File: com.github.emmanueltouzery.cigale.desktop (hunk 1 above)
Comment (emmanueltouzery) [reply com.github.emmanueltouzery.cigale.desktop]: not really. I'll just remove the Comment= line

**Comment by emmanueltouzery** [comment]
bot, build com.github.emmanueltouzery.cigale

**Comment by emmanueltouzery** [comment]
bot, build com.github.emmanueltouzery.cigale

**Comment by emmanueltouzery** [comment]
ok since the files were already in the upstream repo, I just removed it from this repo and changed the JSON to use them.

File: com.github.emmanueltouzery.cigale.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    "--socket=fallback-x11",
+    "--socket=wayland",
+    "--device=dri",
+    "--filesystem=home",
+    "--share=network"
+  ],
+  "build-options": {
+    "append-path": "/usr/lib/sdk/rust-stable/bin",
+    "env": {
+      "CARGO_HOME": "/run/build/rust-flatpak/cargo"
```
Comment (bilelmoussaoui) [inline com.github.emmanueltouzery.cigale.json]:
```suggestion
      "CARGO_HOME": "/run/build/cigale/cargo"
```
With this changed, you can remove the cargo-config dir as it's not needed.

File: update-cargo-sources.sh [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# must have a checkout from the cigale app of the right version in ../cigale
+mkdir .cargo
```
Comment (bilelmoussaoui) [inline update-cargo-sources.sh]: The repository should contain only the files needed to package the app itself. Mind removing this?

Comment (emmanueltouzery) [reply update-cargo-sources.sh]: So a good practice would be to have that script in the main Cigale repo? I'm ok to move it there if it's considered a good practice

File: com.github.emmanueltouzery.cigale.json (hunk 4 above)
Comment (emmanueltouzery) [reply com.github.emmanueltouzery.cigale.json]: I don't fully understand. I'll change that line and check if everything still works afterwards

File: update-cargo-sources.sh (hunk 5 above)
Comment (bilelmoussaoui) [reply update-cargo-sources.sh]: That's up to you really, you can read more about what should/shouldn't be part of the Flathub repository https://github.com/flathub/flathub/wiki/App-Maintenance#required-files

File: com.github.emmanueltouzery.cigale.json (hunk 4 above)
Comment (bilelmoussaoui) [reply com.github.emmanueltouzery.cigale.json]: I'm not sure what you don't understand? that line sets the CARGO HOME to /run/build/$module-name/cargo. In your case, the module name is cigale. Your vendored files can then be found and everything should work without including a custom cargo config file.

**Comment by emmanueltouzery** [comment]
bot, build com.github.emmanueltouzery.cigale

**Comment by emmanueltouzery** [comment]
thank you for the explanations! they were very helpful. and thank you for the patience :-)

**Comment by bilelmoussaoui** [comment]
Happy to help 
/merge

**Comment by bilelmoussaoui** [comment]
@nedrichards it seems the merge action fails to parse the json. I need to switch it json-glib first I guess

**Comment by barthalion** [comment]
It's not about json-glib, looks like `flatpak-builder --show-manifest` prints invalid JSON causing appid detection to fail.

**Comment by barthalion** [comment]
Created the repository with /merge predecessor, I will get back to the issue later.

**Comment by barthalion** [comment]
A repository for this has been created: https://github.com/flathub/com.github.emmanueltouzery.cigale

You will receive an invitation to be a collaborator which will grant you write access to the repository above. The invite can be also viewed [here](https://github.com/flathub/org.shadered.SHADERed/invitations).


If you have never maintained an application before, common questions are answered in [the app maintenance guide](https://github.com/flathub/flathub/wiki/App-Maintenance).


Thanks!

**Pull request closed without merging**